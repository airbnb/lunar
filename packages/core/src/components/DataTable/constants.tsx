import { WidthProperties } from './types';

export const SELECTION_OPTIONS = {
  ACTIVE: 'ACTIVE',
  DISABLED: 'DISABLED',
  INACTIVE: 'INACTIVE',
  HAS_ACTIVE_CHILD: 'HAS_ACTIVE_CHILD',
};

export const STATUS_OPTIONS = {
  ALERT: 'ALERT',
  WARNING: 'WARNING',
};

type HeightMap = {
  [key: string]: number;
};

export const HEIGHT_TO_PX: HeightMap = {
  micro: 32,
  small: 48,
  regular: 56,
  large: 64,
  jumbo: 108,
};

export const DEFAULT_WIDTH_PROPERTIES: WidthProperties = {
  width: 50,
  flexGrow: 1,
  flexShrink: 1,
  minWidth: 0,
};

export const EXPANDABLE_COLUMN_WIDTH = 50;

export const SELECTABLE_COLUMN_WIDTH = 50;
export const SELECTABLE_COLUMN_WIDTH_EXPANDABLE = 39;
