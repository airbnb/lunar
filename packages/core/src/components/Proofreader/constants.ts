import { ARROW_LEFT, ARROW_UP, ARROW_DOWN, ARROW_RIGHT } from '../../keys';

export const AIRBNB_REGEX = /\b(((air|ari|iar)[bn]{3})(?!\.com))\b/gi;
export const NON_WORD_REGEX = /\W/;

export const ARROW_KEYS = [ARROW_LEFT, ARROW_UP, ARROW_DOWN, ARROW_RIGHT];

export const NO_LOCALE = 'none';
export const AUTO_DETECT_LOCALE = 'auto';
export const LOCALE_TO_LT_LOCALE: { [locale: string]: string } = {
  de: 'de-DE',
  en: 'en-US',
  ja: 'ja-JP',
  pt: 'pt-PT',
  ru: 'ru-RU',
  zh: 'zh-CN',
};
