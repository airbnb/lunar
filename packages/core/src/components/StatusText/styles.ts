import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ color }) => ({
  text_danger: {
    color: color.core.danger[4],
  },

  text_info: {
    color: color.core.primary[4],
  },

  text_muted: {
    color: color.muted,
  },

  text_notice: {
    color: color.core.secondary[4],
  },

  text_success: {
    color: color.core.success[4],
  },

  text_warning: {
    color: color.core.warning[5],
  },
});

export default styleSheet;

export { styleSheet };
