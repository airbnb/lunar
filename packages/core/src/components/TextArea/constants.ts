// eslint-disable-next-line import/prefer-default-export
export const LANGUAGE_RULES = {
  UNNECESSARY_ADMIN_NOTE_RULE: 'AGENT_TOOLS_UNNECESSARY_ADMIN_NOTE_RULE',
};
