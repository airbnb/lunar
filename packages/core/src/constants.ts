import { Currency, Locale, LocaleDefinition, TimeZone } from './types';

// I18N/L10N
export const DEFAULT_LOCALE: Locale = 'en';
export const DEFAULT_CURRENCY: Currency = 'USD';
export const DEFAULT_TIMEZONE: TimeZone = 'UTC';

export const LOCALES: Locale[] = [
  'de', // German
  'en', // English
  'es', // Spanish
  'fr', // French
  'it', // Italian
  'ja', // Japanese
  'ko', // Korean
  'ms', // Malay
  'nl', // Dutch
  'pt', // Portuguese
  'ru', // Russian
  'tr', // Turkish
  'zh', // Mandarin
];

// Used for Google Translate API
export const TRANSLATE_LOCALES: LocaleDefinition[] = [
  { locale: 'ar', label: 'ÿπÿ±ÿ®Ÿâ', name: 'Arabic' },
  { locale: 'ca', label: 'Catal√†', name: 'Catalan' },
  { locale: 'cs', label: 'ƒåe≈°tina', name: 'Czech' },
  { locale: 'da', label: 'Dansk', name: 'Danish' },
  { locale: 'de', label: 'Deutsch', name: 'German' },
  { locale: 'el', label: 'ŒïŒªŒªŒ∑ŒΩŒπŒ∫Œ¨', name: 'Greek' },
  { locale: 'en', label: 'English', name: 'English' },
  { locale: 'es', label: 'Espa√±ol', name: 'Spanish' },
  { locale: 'es-419', label: 'Espa√±ol (Am√©rica Latina)', name: 'Spanish (Latin America)' },
  { locale: 'fi', label: 'Suomalainen', name: 'Finnish' },
  { locale: 'fr', label: 'Fran√ßais', name: 'French' },
  { locale: 'he', label: '◊¢÷¥◊ë◊®÷¥◊ô◊™', name: 'Hebrew' },
  { locale: 'hr', label: 'Hrvatski', name: 'Croatian' },
  { locale: 'hu', label: 'Magyar', name: 'Hungarian' },
  { locale: 'id', label: 'Bahasa Indonesia', name: 'Indonesian' },
  { locale: 'is', label: '√çslensku', name: 'Icelandic' },
  { locale: 'it', label: 'Italiano', name: 'Italian' },
  { locale: 'ja', label: 'Êó•Êú¨Ë™û', name: 'Japanese' },
  { locale: 'ko', label: 'ÌïúÍµ≠Ïñ¥', name: 'Korean' },
  { locale: 'ms', label: 'Melayu', name: 'Malay' },
  { locale: 'nl', label: 'Nederlands', name: 'Dutch' },
  { locale: 'no', label: 'Norsk', name: 'Norwegian' },
  { locale: 'pl', label: 'Polskie', name: 'Polish' },
  { locale: 'pt', label: 'Portugu√™s', name: 'Portuguese' },
  { locale: 'pt-PT', label: 'Portugu√™s (Portugal)', name: 'Portuguese (Portugal)' },
  { locale: 'ru', label: '–†—É—Å—Å–∫–∏–π', name: 'Russian' },
  { locale: 'sv', label: 'Svenska', name: 'Swedish' },
  { locale: 'th', label: '‡πÑ‡∏ó‡∏¢', name: 'Thai' },
  { locale: 'tr', label: 'T√ºrk√ße', name: 'Turkish' },
  { locale: 'vi', label: 'Ti·∫øng Vi·ªát', name: 'Vietnamese' },
  { locale: 'zh', label: '‰∏≠Êñá (ÁÆÄ)', name: 'Chinese (Simplified)' },
  { locale: 'zh-TW', label: '‰∏≠Êñá (ÁπÅ)', name: 'Chinese (Traditional)' },
];

// Used for Language Tool (not all are listed)
// https://languagetool.org/http-api/swagger-ui/#!/default/get_languages
export const LT_LOCALES: LocaleDefinition[] = [
  { locale: 'ca-ES', label: 'Catal√†', name: 'Catalan' },
  { locale: 'ca-ES-valencia', label: 'Catal√† (Valenci√†)', name: 'Catalan' },
  { locale: 'da-DK', label: 'Dansk', name: 'Danish' },
  { locale: 'de-AT', label: 'Deutsch (√ñsterreich)', name: 'German (Austria)' },
  { locale: 'de-DE', label: 'Deutsch (Deutschland)', name: 'German (Germany)' },
  { locale: 'el', label: 'ŒïŒªŒªŒ∑ŒΩŒπŒ∫Œ¨', name: 'Greek' },
  { locale: 'en-AU', label: 'English (Australia)' },
  { locale: 'en-CA', label: 'English (Canada)' },
  { locale: 'en-GB', label: 'English (Great Britain)' },
  { locale: 'en-NZ', label: 'English (New Zealand)' },
  { locale: 'en-US', label: 'English (United States)' },
  { locale: 'es', label: 'Espa√±ol (Espa√±a)', name: 'Spanish (Spain)' },
  { locale: 'fr', label: 'Fran√ßais', name: 'French' },
  { locale: 'it', label: 'Italiano', name: 'Italian' },
  { locale: 'ja-JP', label: 'Êó•Êú¨Ë™û', name: 'Japanese' },
  // Korean ÌïúÍµ≠Ïñ¥
  // Malay Melayu
  { locale: 'nl', label: 'Nederlands', name: 'Dutch' },
  { locale: 'pl-PL', label: 'Polskie', name: 'Polish' },
  { locale: 'pt-PT', label: 'Portugu√™s (Portugal)', name: 'Portuguese (Portugal)' },
  { locale: 'pt-BR', label: 'Portugu√™s (Brasil)', name: 'Portuguese (Brazil)' },
  { locale: 'ro-RO', label: 'Rom√¢nƒÉ', name: 'Romanian' },
  { locale: 'ru-RU', label: '–†—É—Å—Å–∫–∏–π', name: 'Russian' },
  { locale: 'sv', label: 'Svenska', name: 'Swedish' },
  // Turkish T√ºrk√ße
  { locale: 'uk-UA', label: '–£–∫—Ä–∞—ó–Ω—Å—å–∫–∞', name: 'Ukrainian' },
  { locale: 'zh-CN', label: '‰∏≠Êñá (ÁÆÄ)', name: 'Chinese (Simplified)' },
];

// DATE/TIME
export const FORMAT_YMD = 'yyyy-MM-dd';
export const FORMAT_PREFIX_DAY = 'ccc'; // Sun
export const FORMAT_SUFFIX_TIMEZONE = 'ZZZZ'; // PST

// MISC
export const MICROS = 1000000;

// THEMES
export const STATUSES = ['notice', 'info', 'success', 'warning', 'danger', 'muted'];
export const BRANDS = ['luxury', 'plus'];
export const Z_INDEX_MODAL = 2000;
export const Z_INDEX_PORTAL = 2000; // Same as Modal to fix Tooltip in Modal
export const Z_INDEX_TOAST = 3000;

// EMOJIS
export const EMOJI_WHITELIST = [
  '1F600', // üòÄ grinning face
  '1F60A', // üòä smiling face with smiling eyes
  '1F642', // üôÇ slightly smiling face
  '1F389', // üéâ party popper
  '2728', // ‚ú® sparkles
  '1F334', // üå¥ palm tree
  '1F333', // üå≥ deciduous tree
  '1F335', // üåµ cactus
  '2600', // ‚òÄÔ∏è sun
  '2744', // ‚ùÑÔ∏è snowflake
  '1F3E0', // üè†Ô∏è house
  '1F697', // üöó automobile
  '2708', // ‚úàÔ∏è airplane
  '1F682', // üöÇ locomotive
  '26F5', // ‚õµÔ∏è sailboat
];

// REACT
export const PASSTHROUGH_WRAPPER_NAMES = ['Tooltip', 'Permission', 'Trebuchet', 'ERF', 'Memo'];
export const STRIP_HOC_NAMES = ['Memo', 'Proxy', 'withStyles', 'withTheme', 'connect'];
