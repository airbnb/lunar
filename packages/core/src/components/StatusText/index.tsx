import React from 'react';
import { mutuallyExclusiveTrueProps } from 'airbnb-prop-types';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import Text, { Props as TextProps } from '../Text';
import { STATUSES } from '../../constants';

const statusPropType = mutuallyExclusiveTrueProps(...STATUSES);

export type Props = TextProps & {
  /** The text to render. */
  children: NonNullable<React.ReactNode>;
  /** Dangerous/failure status (red). */
  danger?: boolean;
  /** Informational status (blue). */
  info?: boolean;
  /** Muted/disabled status (gray). */
  muted?: boolean;
  /** Notice status. */
  notice?: boolean;
  /** Successful status (green). */
  success?: boolean;
  /** Warning status (yellow). */
  warning?: boolean;
};

/** Display a string of classified text with colorful statuses. */
export class StatusText extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    danger: statusPropType,
    info: statusPropType,
    muted: statusPropType,
    notice: statusPropType,
    success: statusPropType,
    warning: statusPropType,
  };

  static defaultProps = {
    danger: false,
    info: false,
    muted: false,
    notice: false,
    success: false,
    warning: false,
  };

  render() {
    const {
      cx,
      children,
      danger,
      info,
      muted,
      notice,
      styles,
      success,
      warning,
      ...restProps
    } = this.props;

    return (
      <Text {...restProps}>
        <span
          className={cx(
            danger && styles.text_danger,
            info && styles.text_info,
            muted && styles.text_muted,
            notice && styles.text_notice,
            success && styles.text_success,
            warning && styles.text_warning,
          )}
        >
          {children}
        </span>
      </Text>
    );
  }
}

export default withStyles(({ color }) => ({
  text_danger: {
    color: color.core.danger[4],
  },

  text_info: {
    color: color.core.primary[4],
  },

  text_muted: {
    color: color.muted,
  },

  text_notice: {
    color: color.core.secondary[4],
  },

  text_success: {
    color: color.core.success[4],
  },

  text_warning: {
    color: color.core.warning[5],
  },
}))(StatusText);
