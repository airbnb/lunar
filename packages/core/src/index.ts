import Aesthetic, { FontFace } from 'aesthetic';
import AphroditeAesthetic, { NativeBlock, ParsedBlock } from 'aesthetic-adapter-aphrodite';
import { Settings as LuxonSettings } from 'luxon';
import { Path as EmojiPath } from 'interweave-emoji';
import globalStyles from './themes/global';
import lightTheme from './themes/light';
import darkTheme from './themes/dark';
import getLocaleFromClient from './utils/getLocaleFromClient';
import getTimezoneFromClient from './utils/getTimezoneFromClient';
import { DEFAULT_LOCALE, DEFAULT_TIMEZONE } from './constants';
import {
  Locale,
  TimeZone,
  Logger,
  Translator,
  TranslateParams,
  TranslateProps,
  Theme,
} from './types';

export type Settings = {
  defaultLocale?: Locale;
  defaultTimezone?: TimeZone;
  emojiCDN?: EmojiPath;
  errorURL?: string;
  fontFaces?: { [fontFamily: string]: FontFace[] };
  fontFamily?: string;
  logger?: Logger | null;
  name: string;
  theme?: 'light' | 'dark';
  translator?: Translator | null;
  translatorComponent?: React.ComponentType<TranslateProps> | null;
};

class Core {
  settings: Required<Settings> = {
    defaultLocale: getLocaleFromClient() || DEFAULT_LOCALE,
    defaultTimezone: getTimezoneFromClient() || DEFAULT_TIMEZONE,
    emojiCDN: '',
    errorURL: '',
    fontFaces: {},
    fontFamily:
      '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
    logger: null,
    name: 'Lunar',
    theme: 'light',
    translator: null,
    translatorComponent: null,
  };

  protected aesthetic?: Aesthetic<Theme, NativeBlock, ParsedBlock>;

  initialize(settings: Settings) {
    this.settings = {
      ...this.settings,
      ...settings,
    };

    this.bootstrapAesthetic();
    this.bootstrapLuxon();
  }

  bootstrapAesthetic() {
    const { theme, fontFaces } = this.settings;
    const fontFamily = this.fontFamily();
    const globals = globalStyles(fontFaces);

    this.aesthetic = new AphroditeAesthetic<Theme>([], {
      theme,
      passThemeProp: false,
      pure: true,
    })
      .registerTheme('light', lightTheme(fontFamily), globals)
      .registerTheme('dark', darkTheme(fontFamily), globals);
  }

  bootstrapLuxon() {
    LuxonSettings.defaultLocale = this.locale();
    LuxonSettings.defaultZoneName = this.timezone();
    LuxonSettings.throwOnInvalid = true;
  }

  getAesthetic() {
    if (__DEV__) {
      if (!this.aesthetic) {
        throw new Error(
          'Aesthetic has not been initialized. Please call `Core.initialize()` from `@airbnb/lunar`.',
        );
      }
    }

    return this.aesthetic!;
  }

  fontFamily(): string {
    const locale = this.locale();

    if (locale.startsWith('ja')) {
      return '"ヒラギノ角ゴ Pro", "Hiragino Kaku Gothic Pro", メイリオ, Meiryo, Osaka, "ＭＳ Ｐゴシック", "MS PGothic", "MS Gothic", "ＭＳ ゴシック", "Helvetica Neue", Helvetica, Arial, sans-serif';
    }

    if (locale.startsWith('ko')) {
      return '"나눔 고딕", "Nanum Gothic", "맑은 고딕", "Malgun Gothic", "Apple Gothic", 돋움, Dotum, "Helvetica Neue", Helvetica, Arial, sans-serif';
    }

    if (locale.includes('zh')) {
      return '"Hiragino Sans GB", 华文细黑, STHeiti, 微软雅黑, "Microsoft YaHei", SimHei, "Helvetica Neue", Helvetica, Arial, sans-serif';
    }

    return this.settings.fontFamily;
  }

  locale(): Locale {
    return this.settings.defaultLocale || DEFAULT_LOCALE;
  }

  log = (error: Error, extra: object) => {
    const { logger } = this.settings;

    if (logger) {
      logger(error, extra);
    }
  };

  timezone(): TimeZone {
    return this.settings.defaultTimezone || DEFAULT_TIMEZONE;
  }

  translate = (phrase: string, params: TranslateParams, context: string) => {
    const { translator } = this.settings;

    if (translator) {
      return translator(phrase, params, context);
    }

    let message = phrase;

    if (phrase.includes('||||')) {
      const [singular, plural] = phrase.split('||||');
      const count = params.smartCount || 0;

      message = count === 1 ? singular : plural;
    }

    // Low-level token interpolation
    return message.replace(/%\{(\w+)\}/g, (match, key) => String(params[key] || ''));
  };
}

export default new Core();
