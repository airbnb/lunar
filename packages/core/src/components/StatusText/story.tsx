import React from 'react';
import LoremIpsum from ':storybook/components/LoremIpsum';
import StatusText from '.';

export default {
  title: 'Core/StatusText',
  parameters: {
    inspectComponents: [StatusText],
  },
};

export function defaultAndStatusClassifiedText() {
  return (
    <>
      <StatusText>
        <LoremIpsum short />
      </StatusText>
      <StatusText notice>
        <LoremIpsum short />
      </StatusText>
      <StatusText info>
        <LoremIpsum short />
      </StatusText>
      <StatusText success>
        <LoremIpsum short />
      </StatusText>
      <StatusText warning>
        <LoremIpsum short />
      </StatusText>
      <StatusText danger>
        <LoremIpsum short />
      </StatusText>
      <StatusText muted>
        <LoremIpsum short />
      </StatusText>
    </>
  );
}

defaultAndStatusClassifiedText.story = {
  name: 'Default and status classified text.',
};

export function canPassPropsToTheUnderlyingComponent() {
  return (
    <>
      <StatusText notice micro>
        <LoremIpsum short />
      </StatusText>
      <StatusText info small>
        <LoremIpsum short />
      </StatusText>
      <StatusText success large>
        <LoremIpsum short />
      </StatusText>
      <StatusText warning disabled>
        <LoremIpsum short />
      </StatusText>
      <StatusText danger light>
        <LoremIpsum short />
      </StatusText>
      <StatusText muted bold>
        <LoremIpsum short />
      </StatusText>
    </>
  );
}

canPassPropsToTheUnderlyingComponent.story = {
  name: 'Can pass props to the underlying component.',
};
