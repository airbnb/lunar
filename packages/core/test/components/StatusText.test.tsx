import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import StatusText from '../../src/components/StatusText';
import { STATUSES } from '../../src/constants';

describe('<StatusText />', () => {
  it('errors when multiple states are used at once', () => {
    expect(() => {
      shallowWithStyles(
        <StatusText danger success>
          Default
        </StatusText>,
      );
    }).toThrowError();
  });

  describe('statuses', () => {
    STATUSES.forEach(status => {
      it('renders text', () => {
        const wrapper = shallowWithStyles(
          shallowWithStyles(<StatusText {...{ [status]: true }}>{status}</StatusText>).getElement(),
        );

        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  it('can pass props to underlying `Text`', () => {
    const wrapper = shallowWithStyles(
      shallowWithStyles(
        <StatusText large muted>
          Text
        </StatusText>,
      ).getElement(),
    );

    expect(wrapper).toMatchSnapshot();
  });
});
