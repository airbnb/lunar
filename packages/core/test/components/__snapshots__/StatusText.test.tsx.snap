// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StatusText /> can pass props to underlying \`Text\` 1`] = `
<h4
  className="text_8h9m6q-o_O-text_large_1aiv2f"
>
  <span
    className="text_muted_10k87om"
  >
    Text
  </span>
</h4>
`;

exports[`<StatusText /> statuses renders text 1`] = `
<div
  className="text_8h9m6q"
>
  <span
    className="text_notice_16fnaxc"
  >
    notice
  </span>
</div>
`;

exports[`<StatusText /> statuses renders text 2`] = `
<div
  className="text_8h9m6q"
>
  <span
    className="text_info_1qlg6bk"
  >
    info
  </span>
</div>
`;

exports[`<StatusText /> statuses renders text 3`] = `
<div
  className="text_8h9m6q"
>
  <span
    className="text_success_xr5ae0"
  >
    success
  </span>
</div>
`;

exports[`<StatusText /> statuses renders text 4`] = `
<div
  className="text_8h9m6q"
>
  <span
    className="text_warning_1b6zul9"
  >
    warning
  </span>
</div>
`;

exports[`<StatusText /> statuses renders text 5`] = `
<div
  className="text_8h9m6q"
>
  <span
    className="text_danger_1ardfxt"
  >
    danger
  </span>
</div>
`;

exports[`<StatusText /> statuses renders text 6`] = `
<div
  className="text_8h9m6q"
>
  <span
    className="text_muted_10k87om"
  >
    muted
  </span>
</div>
`;
